packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -1991,8 +2653,276 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
 
   String _lastAnnouncedRouteName;
 
+  bool _debugUpdatingPage = false;
+  void _updatePages() {
+    assert(() {
+      assert(!_debugUpdatingPage);
+      _debugCheckDuplicatedPageKeys();
+      _debugUpdatingPage = true;
+      return true;
+    }());
+
+    // This attempts to diff the new pages list (widget.pages) with
+    // the old _RouteEntry[s] list (_history), and produce a new list of
+    // _RouteEntry[s] to be the new list of _history. This method roughly
+    // follows the same outline of RenderObjectElement.updateChildren.
+
+    // The cases it tries to optimize for are:
+    //  - the old list is empty
+    //  - All the pages in the new list can match the page-based routes in the old
+    //    list, and their orders are the same.
+    //  - there is an insertion or removal of one or more page-based route in
+    //    only one place in the list
+    // If a page-based route with a key is in both lists, it will be synced.
+    // Page-based routes without keys might be synced but there is no guarantee.
+
+    // The general approach is to sync the entire new list backwards, as follows:
+    // 1. Walk the lists from the top, syncing nodes, and records pageless routes,
+    //    until you no longer have matching nodes.
+    // 2. Walk the lists from the bottom, without syncing nodes, until you no
+    //    longer have matching nodes. We'll sync these nodes at the end. We
+    //    don't sync them now because we want to sync all the nodes in order
```

Comment:
Why do we need to sync them in order?

Reply:
I wrote this document other way around... top should be bottom and bottom should be top.

the second step walk the lists from the top in reverse order. If we sync them now, it will be stored in reverse order as well. we have to reverse them again to merge the results back to new history. It will be O(2n) anyway. however, if we sync them now, we will have to spend another O(n) space to store the reversed list and it does not worth.